File: app/node/color/displaytransform/displaytransform.cpp
/***

  Olive - Non-Linear Video Editor
  Copyright (C) 2022 Olive Team
  Modifications Copyright (C) 2025 mikesolar

  This program is free software: you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.

  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program.  If not, see <http://www.gnu.org/licenses/>.

***/

#include "displaytransform.h"

#include "node/color/colormanager/colormanager.h"

namespace olive
{

const QString DisplayTransformNode::k_display_input =
	QStringLiteral("display_in");
const QString DisplayTransformNode::k_view_input = QStringLiteral("view_in");
const QString DisplayTransformNode::k_direction_input = QStringLiteral("dir_in");

#define super OCIOBaseNode

DisplayTransformNode::DisplayTransformNode()
{
	add_input(k_display_input, NodeValue::k_combo, 0,
			 InputFlags(k_input_flag_not_keyframable | k_input_flag_not_connectable));

	add_input(k_view_input, NodeValue::k_combo, 0,
			 InputFlags(k_input_flag_not_keyframable | k_input_flag_not_connectable));

	add_input(k_direction_input, NodeValue::k_combo, 0,
			 InputFlags(k_input_flag_not_keyframable | k_input_flag_not_connectable));
}

QString DisplayTransformNode::name() const
{
	return tr("Display Transform");
}

QString DisplayTransformNode::id() const
{
	return QStringLiteral("org.olivevideoeditor.Olive.displaytransform");
}

QVector<Node::CategoryID> DisplayTransformNode::category() const
{
	return { k_category_color };
}

QString DisplayTransformNode::description() const
{
	return tr("Converts an image to or from a display color space.");
}

void DisplayTransformNode::retranslate()
{
	super::retranslate();

	set_input_name(k_texture_input, tr("Input"));
	set_input_name(k_display_input, tr("Display"));
	set_input_name(k_view_input, tr("View"));
	set_input_name(k_direction_input, tr("Direction"));
	set_combo_box_strings(k_direction_input, { tr("Forward"), tr("Inverse") });
}

void DisplayTransformNode::InputValueChangedEvent(const QString &input,
												  int element)
{
	Q_UNUSED(element);
	if (input == k_display_input || input == k_direction_input ||
		input == k_view_input) {
		if (input == k_display_input) {
			update_views();
		}
		generate_processor();
	}
}

QString DisplayTransformNode::get_display() const
{
	if (manager()) {
		int index = get_standard_value(k_display_input).toInt();
		if (index < manager()->list_available_displays().size()) {
			return manager()->list_available_displays().at(index);
		}
	}
	return QString();
}

QString DisplayTransformNode::get_view() const
{
	if (manager()) {
		QString display = get_display();
		if (!display.isEmpty()) {
			int index = get_standard_value(k_view_input).toInt();
			QStringList views = manager()->list_available_views(display);
			if (index < views.size()) {
				return views.at(index);
			}
		}
	}
	return QString();
}

ColorProcessor::Direction DisplayTransformNode::get_direction() const
{
	return static_cast<ColorProcessor::Direction>(
		get_standard_value(k_direction_input).toInt());
	;
}

void DisplayTransformNode::update_displays()
{
	if (manager()) {
		set_combo_box_strings(k_display_input, manager()->list_available_displays());
	}
}

void DisplayTransformNode::update_views()
{
	if (manager()) {
		set_combo_box_strings(k_view_input,
						   manager()->list_available_views(get_display()));
	}
}

void DisplayTransformNode::config_changed()
{
	update_displays();
	update_views();
	generate_processor();
}

void DisplayTransformNode::generate_processor()
{
	if (manager()) {
		ColorTransform transform(get_display(), get_view(), QString());
		set_processor(ColorProcessor::create(
			manager(), manager()->get_reference_color_space(), transform,
			get_direction()));
	}
}

}

File: src/node/rust/src/nodes/displaytransform.rs
// Oak Video Editor - Non-Linear Video Editor
// Copyright (C) 2026 Oak Team
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

//! OCIO display transform node (C++
//! `src/node/src/color/displaytransform/displaytransform.{h,cpp}`,
//! `olive::DisplayTransformNode`).
//!
//! Note: OpenColorIO itself is never linked here; it is reached through
//! the color manager (`crate::colormanager`) and the oakrender bridge
//! (`crate::bridge::render`), like the C++ node's
//! `oaknode_colormanager_*` / `oakrender_color_processor_*` calls.

use crate::factory::NodeMeta;
use crate::node::{Category, NodeBehavior, NodeCore};

use super::ociobase::OcioBase;

/// Display combo input id (C++ `k_display_input`). Type: combo; default
/// `0`; flags: not-keyframable, not-connectable. Combo strings are the
/// color manager's available displays (refreshed on config change).
pub const DISPLAY_INPUT: &str = "display_in";

/// View combo input id (C++ `k_view_input`). Type: combo; default `0`;
/// flags: not-keyframable, not-connectable. Combo strings are the views
/// available for the selected display.
pub const VIEW_INPUT: &str = "view_in";

/// Direction combo input id (C++ `k_direction_input`). Type: combo;
/// default `0` (forward); flags: not-keyframable, not-connectable.
/// Combo strings (set in `retranslate`): "Forward", "Inverse".
pub const DIRECTION_INPUT: &str = "dir_in";

/// Display transform node. Converts an image to or from a display color
/// space via an OCIO display/view transform. Owns no members beyond the
/// embedded OCIO base state (C++ has no own private members).
pub struct DisplayTransformNode {
	/// Shared OCIO base state (C++ base class `OCIOBaseNode`).
	base: OcioBase,
}

impl DisplayTransformNode {
	/// Selected display name (C++ `get_display()`): the display combo
	/// index mapped through the color manager's display list; empty
	/// string when no manager is attached or the index is out of range.
	fn get_display(&self, core: &NodeCore) -> String {
		todo!()
	}

	/// Selected view name (C++ `get_view()`): the view combo index
	/// mapped through the manager's views for [`Self::get_display`];
	/// empty when unavailable.
	fn get_view(&self, core: &NodeCore) -> String {
		todo!()
	}

	/// Transform direction (C++ `get_direction()`): the direction combo
	/// value cast to `ColorProcessor::Direction` (`0` = normal/forward,
	/// `1` = inverse).
	fn get_direction(&self, core: &NodeCore) -> i64 {
		todo!()
	}

	/// Refresh the display combo strings from the manager (C++
	/// `update_displays()`); no-op without a manager.
	fn update_displays(&mut self, core: &mut NodeCore) {
		todo!()
	}

	/// Refresh the view combo strings for the current display (C++
	/// `update_views()`); no-op without a manager.
	fn update_views(&mut self, core: &mut NodeCore) {
		todo!()
	}

	/// (Re)build the color processor (C++ `generate_processor()`): wraps
	/// the manager, builds a display transform for
	/// display/view/reference-space with the selected direction via
	/// `oakrender_color_processor_create_transform`, and stores it with
	/// [`OcioBase::set_processor`].
	fn generate_processor(&mut self, core: &mut NodeCore) {
		todo!()
	}

	/// OCIO config change hook (C++ `config_changed()` override):
	/// refreshes displays and views, then regenerates the processor.
	fn config_changed(&mut self, core: &mut NodeCore) {
		todo!()
	}
}

impl NodeBehavior for DisplayTransformNode {
	/// Human-readable name (C++ `name()`).
	fn name(&self) -> &str {
		"Display Transform"
	}

	/// Stable type id (C++ `id()`).
	fn type_id(&self) -> &str {
		"org.olivevideoeditor.Olive.displaytransform"
	}

	/// Categories (C++ `category()`).
	fn categories(&self) -> &[Category] {
		&[Category::Color]
	}

	/// Description (C++ `description()`).
	fn description(&self) -> &str {
		"Converts an image to or from a display color space."
	}

	/// Localized input names (C++ `retranslate()`): `tex_in` -> "Input",
	/// `display_in` -> "Display", `view_in` -> "View", `dir_in` ->
	/// "Direction" (also sets the direction combo strings
	/// "Forward"/"Inverse").
	fn input_name<'a>(&self, id: &'a str) -> &'a str {
		todo!()
	}

	/// Input value changed (C++ `InputValueChangedEvent`): for
	/// `display_in`, `view_in` or `dir_in` regenerates the processor;
	/// a `display_in` change additionally refreshes the view combo.
	fn input_value_changed(&mut self, core: &mut NodeCore, input: &str, element: i32) {
		todo!()
	}

	/// Evaluate outputs: inherited from the C++ base
	/// (`OCIOBaseNode::value()`), i.e. delegates to
	/// [`OcioBase::value`] — color-transform job when the processor is
	/// ready, pass-through otherwise.
	fn value(
		&self,
		core: &NodeCore,
		inputs: &crate::value::NodeValueRow,
		time: oakcore_rs::Rational,
		table: &mut crate::value::NodeValueTable,
	) {
		todo!()
	}

	/// Added to a graph (C++ base `AddedToGraphEvent`): captures the
	/// project's color manager and runs `config_changed()` via
	/// [`OcioBase::added_to_graph`].
	fn added_to_graph(&mut self, core: &mut NodeCore) {
		todo!()
	}

	/// Removed from a graph (C++ base `RemovedFromGraphEvent`): clears
	/// the color manager pointer via [`OcioBase::removed_from_graph`].
	fn removed_from_graph(&mut self, core: &mut NodeCore) {
		todo!()
	}

	/// Deep copy (C++ `copy()` via `NODE_DEFAULT_FUNCTIONS`).
	fn duplicate(&self, core: &NodeCore) -> Option<Box<dyn NodeBehavior>> {
		todo!()
	}
}

/// Constructor (C++ `DisplayTransformNode::DisplayTransformNode()`):
/// builds the base (`tex_in` texture input, effect input, video-effect
/// flag) and adds the `display_in`/`view_in`/`dir_in` combo inputs with
/// the defaults and flags documented on the constants.
pub fn create() -> (NodeCore, Box<dyn NodeBehavior>) {
	todo!()
}

/// Register this node type (C++ factory entry for
/// `org.olivevideoeditor.Olive.displaytransform`).
pub fn register(meta: &mut Vec<NodeMeta>) {
	meta.push(NodeMeta {
		type_id: "org.olivevideoeditor.Olive.displaytransform",
		name: "Display Transform",
		categories: &[Category::Color],
		create,
	});
}
